controller/lib/core/nvparam.cpp

```diff
@@ -0,0 +1,218 @@
+// it.  In this way there should always be at least one valid copy of the
+// parameters in flash memory and normally only one.
+
+#include "nvparam.h"
+#include "checksum.h"
+#include "flash.h"
+#include "hal.h"
+#include <string.h>
+
+// Size of the parameter block including the header
+static constexpr uint32_t nvparam_size = 512;
+
+// This is the value of the 'mark' field of a good structure
+// in flash.  It's pretty arbitrary, anything but 0 or 0xff
+static constexpr uint8_t good_mark = 0x55;
+
+// Address of valid parameter block
+static uint32_t nvparam_addr;
+
+// local functions
+static bool CheckStruct(uint32_t addr);
+static uint32_t BlockCRC(void *blk);
+static void Invalidate(uint32_t addr);
+static bool SaveBlock(NVparams *blk, uint32_t addr);
+
+// One time init of non-volatile parameter area.
+void NVparamsInit(void) {
+  // Make sure the size of the storage structure is correct.
+  // This catches mistakes when adding new variables.
+  // If it's wrong, then I call a function that doesn't actually
+  // exist.  That generates a link time error, so the program
+  // won't actually be built.
+  // If the size is correct then the compiler will optimize this out.
+  if (sizeof(NVparams) != nvparam_size) {
+    extern void CheckSizeOfNVparams(void);
+    CheckSizeOfNVparams();
+  }
+
+  // Total structures in two pages of flash
+  int N = flash_params_size / nvparam_size;
+
+  // Run through structures looking for a valid one.
+  uint32_t addr = flash_params_start;
+  for (int i = 0; i < N; i++, addr += nvparam_size) {
+    // Check the structure at this address
+    if (!CheckStruct(addr))
+      continue;
+
+    // If this is the first valid block we've found,
+    // just keep track of the address
+    if (!nvparam_addr) {
+      nvparam_addr = addr;
+      continue;
+    }
+
+    // I've found multiple valid blocks.  This normally shouldn't
+    // happen but could if power is lost at just the right time
+    // when updating flash.
+    //
+    // I'll check the counter to see which was is newer and keep
+    // that address.  I'll also invalidate the older one.
+    NVparams *a = reinterpret_cast<NVparams *>(nvparam_addr);
+    NVparams *b = reinterpret_cast<NVparams *>(addr);
+
+    int8_t diff = static_cast<int8_t>(b->count - a->count);
+    if (diff > 0) {
+      Invalidate(nvparam_addr);
+      nvparam_addr = addr;
+    } else
+      Invalidate(addr);
+  }
+
+  // If I found a valid parameter area, I'm done
+  if (nvparam_addr)
+    return;
+
+  // If no valid storage was found then I'll just create one with
+  // zero values.
+  NVparams blank;
+  memset(&blank, 0, sizeof(blank));
+  blank.mark = good_mark;
+  blank.crc = BlockCRC(&blank);
+
+  Hal.FlashErasePage(flash_params_start);
+  Hal.FlashWrite(flash_params_start, reinterpret_cast<uint32_t *>(&blank),
+                 sizeof(blank) / sizeof(uint32_t));
+  nvparam_addr = flash_params_start;
+}
+
+const NVparams *FindStore(void) {
```

Comment:
I'm finding this name kind of misleading, because it doesn't actually go and find the store.

`NVparamsGet()`?  `GetNVparams()`?